### ISSUE TITLE
v0.6rc2 improvements/clarifications

### ISSUE THREAD

**Issue by afrendeiro** [epigen/looper#144]
Hi Vince and Nathan,
I had a go with v0.6rc2 and since I encountered several hiccups (now in retrospect I think mostly caused by old docs), I just to be sure none of the issues was caused by me and I just took some time to go over them.
My comments follow:
- Logging:
    - Could there be an easy way to let the user see some of the logging options in the CLI without cluttering too much the `--help` message? Right now one has to look at the code to activate more verbose logging.
    - ~~Similarly, could we add Vince's way of logging in interactive section (which I somehow cannot find anywhere in the pull/issue discussions) to the docs?~~ #131 <- already an issue
    - the `--dbg` option doesn't seem to do anything for me, I still get the same amount of detail. When using the interactive session however (as described in #129), I do get more detail.
    - [really minor thing: can the argparse arguments that receive values (e.g. [--logfile](https://github.com/epigen/looper/blob/master/looper/looper.py#L59) ) have a `dest=` keyword making explicit the destination variable? This makes it easy to track where certain variables come from/are used again.]
- Sample subclasses <small>(this took me quite a while to understand - some of the comments actually don't apply, but since I wrote them as I explored, they reflect my steps in trying to understanding how this works and can be useful to make a few things more explicit/documented)</small>:
    - How is this implemented exactly now? [This section of the docs](http://looper.readthedocs.io/en/latest/models.html#extending-sample-objects) probably needs to be updated. I couldn't figure out how to let looper know about my classes.
    - Importantly, this should be independent of looper and just handled by `models`. Not entirely sure if it is like that right now the flow now is `build_submission_bundles` -> `fetch_sample_subtype` -> `_import_sample_subtype` but `build_submission_bundles` is only called by looper and not somewhere at Project initialization as I'd have thought. In fact there is no logic in Project to accommodate sample subclasses, which now only uses `_make_basic_samples`.
    - I got some confirmation of the above when I saw the following message in my looper log: "'/home/arendeiro/workspace/pipelines/pipelines/rnaBitSeq.py' appears to attempt to run on import; does it lack a conditional on '__main__'? Using base type: Sample" <- so it is trying to use subclasses! (This pipeline in particular does not use classes and does not extend Sample though). When I inspected my Yaml files of another data type/pipeline, my additional attributes were there! Maybe the logger should be more refined here is saying something like "Using subclass X for sample A" or maybe just my logging from CLI doesn't work.
    - But despite working, this means that Sample subclasses are only triggered by looper and are not present in the samples of the Project. This would be really useful for the future use of models outside of looper! (and my current use too) - can we please have this implemented in this way?
- Sample-Project interaction/link:
    - Can we reverse the removal of Sample-Project linking? It is true these don't need to exist, and probably shouldn't be used in the looper workflow but are extremely useful outside of it.

TODO:
- [x] Logging CLI help (#145)
- [x] Explicit CLI option names (#146)
- [x] Document Sample subclasses (#147)
- [x] Logging options docs (#148)
- [x] relink samples/projects (for now -- to be revisited later) (#150)

**Comment by nsheff** [comment]
Ok, lots of stuff here.

>Logging:
Could there be an easy way to let the user see some of the logging options in the CLI without cluttering too much the --help message? Right now one has to look at the code to activate more verbose logging.

I propose: a detailed documentation page in readthedocs, with a note in the CLI help message.

> the --dbg option doesn't seem to do anything for me, I still get the same amount of detail. When using the interactive session however (as described in #129), I do get more detail.

make sure you put `--dbg` right after `looper` (so, `looper --dbg` NOT `looper run config.yaml --dbg`) because this parameter is parsed _before_ the command. It's a python thing. Any suggestions for documenting this better?

> [really minor thing: can the argparse arguments that receive values (e.g. --logfile ) have a dest= keyword making explicit the destination variable? This makes it easy to track where certain variables come from/are used again.]

I agree with you on this one, but @vreuter does not :stuck_out_tongue: 

>How is this implemented exactly now? This section of the docs probably needs to be updated. I couldn't figure out how to let looper know about my classes.

Yeah. sorry about that. we didn't write any docs for it yet. Oops. The only place it's discussed is in issue 130. it's quite simple, you just add a line to the pipeline interface -- and this is actually _not needed unless you have more than 1 sample subtype_. If you specify only a single child class that extends `Sample`, looper will automatically find it and use it.

> Importantly, this should be independent of looper and just handled by models. Not entirely sure if it is like that right now the flow now is build_submission_bundles -> fetch_sample_subtype -> _import_sample_subtype but build_submission_bundles is only called by looper and not somewhere at Project initialization as I'd have thought. In fact there is no logic in Project to accommodate sample subclasses, which now only uses _make_basic_samples.
I got some confirmation of the above when I saw the following message in my looper log: "'/home/arendeiro/workspace/pipelines/pipelines/rnaBitSeq.py' appears to attempt to run on import; does it lack a conditional on 'main'? Using base type: Sample" <- so it is trying to use subclasses! (This pipeline in particular does not use classes and does not extend Sample though). When I inspected my Yaml files of another data type/pipeline, my additional attributes were there! Maybe the logger should be more refined here is saying something like "Using subclass X for sample A" or maybe just my logging from CLI doesn't work.
But despite working, this means that Sample subclasses are only triggered by looper and are not present in the samples of the Project. This would be really useful for the future use of models outside of looper! (and my current use too) - can we please have this implemented in this way?

well -- good points. key thing to remember: a Sample subclass is_not_ inherent to a sample, but is relative to a sample-pipeline combination. Different pipelines could specify different subtypes for a given input sample. So, if you want sample subclasses, you need a pipeline interface. In other words it will **not** be handled by Project -- but it is handled by the pipeline interface classes, which are also in `models`. It should be possible to do what you're saying while still maintaining the child subtypes as relative to a pipeline... if only 1 pipeline is specified in a pipeline interface it should default to that child class, but if more are present, it must default to a basic class unless a particular subclass is specified somehow. This may take a bit more thought.

> Sample-Project interaction/link:
> Can we reverse the removal of Sample-Project linking? It is true these don't need to exist, and probably shouldn't be used in the looper workflow but are extremely useful outside of it.

I propose: make the samples project-aware, but don't use this anywhere in the code (so that the dev can use it outside models).

**Comment by vreuter** [comment]
I'll address the other points separately, but I want to first focus on the samples/project deal. Why is that good/desirable? I see value in a sample knowing about certain attributes of a project that are sample-independent, but it would be bizarre for a sample to know about other samples via a project, especially with inter-sample independence advertised as a bedrock looper principle. I think it'd be good to facilitate access by each sample to a subset of project metadata, but I strongly oppose the existence of a direct reference of a sample to its project. Imports, function calls, object references -- I think circular is bad.

**Comment by nsheff** [comment]
Sample/Project link: I see both sides. I don't have strong feelings at this point. I think if you have a sample, you probably got it through a project, so why do you need the pointer? Just remember where you pulled the sample. But @afrendeiro if you can give a specific use case for when you'd have a Sample, need it's Project, and not have accessed the sample through that Project in the first place (necessitating a link to Project from Sample), that would be interesting.

**Comment by vreuter** [comment]
LOGGING

TLDR -- I like the idea of reserving detail for docs but a note in the CLI about the logging options, and I can hook up the debug/more detailed stuff to the interactive session/module. @nsheff apparently knows me too well; I think the ease of renaming offered by using the same option name on the user and internal sides is nice, and so then I just omit `dest`, but if you both want it, then I'm fine with adding it.

More thorough -- What would be the mechanism for the note for the logging? An addendum to the text passed as the argument to the `epilog` parameter in the argument parser? I like that idea, and then there we could also note the positional requirement for the `--dbg`. Otherwise, I think we'd have to not `argparse.SUPPRESS` the `--dbg` help text and explain it there. I agree with @afrendeiro that the specific placement of the `--dbg` is mildly annoying--and it's tripped me up a couple of times before I remember this. I think that it'd be nice to eventually see if there may be a way to remove that requirement through a little sparring with `argparse`, but for now it sounds good to just make note of it in the CLI.

**Comment by nsheff** [comment]
> What would be the mechanism for the note for the logging? An addendum to the text passed as the argument to the epilog parameter in the argument parser? 

Sure, something at the end -- I think something simple like: `For advanced options like increasing debug output verbosity and changing logging message formats, please see the documentation` -- or better yet, could there be a `--adv-help` command that describes the advanced help stuff? I like that!

Then we add a new page for usage that would have an 'advanced usage' section. And this should be automatic. As outlined in #73

**Comment by nsheff** [comment]
this would be really great; then we only maintain the docs right in place and then auto-generate the readthe docs; it's suppressed by default but accessible on the CLI if someone wants. keeps general use help uncluttered. I think this solves all issues. name ideas? `--adv-help` ? Meh...

**Comment by nsheff** [comment]
> SAMPLE SUBCLASSES ...

Great explanation. This post is the beginning of the documentation for this...

**Comment by afrendeiro** [comment]
> make sure you put --dbg right after looper (so, looper --dbg NOT looper run config.yaml --dbg) 
\+
> I agree with @afrendeiro that the specific placement of the --dbg is mildly annoying--and it's tripped me up a couple of times before

Thanks for pointing it out. This is a bit annoying, but completely fine if you do give the user the chance to know that there is a `--dbg` option and its placement by seeing that it only appears when you do `looper --help` and doesn't when you do `looper run --help`.

> But @afrendeiro if you can give a specific use case for when you'd have a Sample, need its Project, and not have accessed the sample through that Project in the first place (necessitating a link to Project from Sample), that would be interesting.

Well I generally don't ask for features if I'm not using them:
My case is in when running a pipeline. Imagine your pipeline could make use of a few project-centric variables like a specific directory where you want to group a certain type of output for all samples (e.g. bigwig files, some pdfs, etc - this is not the norm for all of the sample's output but just one or two). This variable is common to all samples of the project and could simply be defined by extending the project's config with it. Instead of always having this passed as a parameter, one could simply read the sample's config (which is the only thing I require as input) and get that variable though the project's.
I do agree though that there must be a way to contain circularity and recursiveness specially in the serialization of the objects to yaml (I believe something was/is in place to assure this), so yes it is not straightforward to do this out of the box - it would be good if possible though.

> SAMPLE SUBCLASSES

Also agree with everything.
By the way, this also accomodates further class inheritance too: 
```
ChIPSeqSample(Sample)
ChIPmentationSample(ChIPSeqSample)
```

**Comment by vreuter** [comment]
Thanks @afrendeiro  for pointing out the grandchild-and-beyond subclassing ability available; I'll add this to the docs. The way that I've written the subclasses section of the docs in https://github.com/epigen/looper/pull/149 doesn't suggest that it isn't possible, but it also doesn't drawn attention to this capability, and I think that'd be good to do.

**Comment by vreuter** [comment]
Intuitively, it seems like the portion of a `Project` that's most useful and logical to share among / distribute to each `Sample` is the `metadata` section, and it sounds like the directory scenario that you're describing would fit into this framework; is that right? If so, would it be acceptable to allow each `Sample` to access the `metadata` section components rather than the entire `Project` object. I think that could fit nicely with the mental model of what's going on with the objects and avoid circularity concerns, so if it will meet usage needs, that's what I'd like to implement.

**Comment by nsheff** [comment]
> would it be acceptable to allow each Sample to access the metadata section components rather than the entire Project object.

@vreuter  I would suggest either an all-or-nothing here. let's not add the complexity of figuring out what it can or can't do... this is a slippery slope and I think it's not worth our time to go down that route. either it's useful for a sample to know its project, or it's not. let's keep it simple...

>Instead of always having this passed as a parameter, one could simply read the sample's config (which is the only thing I require as input) and get that variable though the project's.

@afrendeiro could you instead add this variable to the sample object itself at object creation?

**Comment by vreuter** [comment]
Agreed, I think we're suggesting almost the same thing. That said, I do think there's a major anti-complexity win in that small difference as you alluded to. Attaching the raw paths that would be in metadata directly to the Sample instead--as simple string-valued attributes and accession them directly rather than through metadata--would achieve this, so I'm all for it!

**Comment by nsheff** [comment]
I think for now, I don't want to delay this release any further and I think we still should think further about this. So I think we should add the link back in (just add a pointer at `Sample.project`) so @afrendeiro can continue to use it in his use case. Let's think about this and continue to discuss and then revisit this in a few months at version 0.7. I'd like to finish off 0.6 which I feel is already overdue.

**Comment by nsheff** [comment]
Ok let's continue the sample/project discussion in #150 and @vreuter can you just set a simple link for now to close this master-ticket?

**Comment by vreuter** [comment]
For `Sample`-to-`Project` link, I've implemented a workaround that includes the `metadata` section. Inclusion of the entire `Project` reignited the circularity/recursion issues during serialization that I experienced amid the removal of the link a couple weeks ago and that we anticipated in this discussion. This felt acceptable given the use case discussed and the previously mentioned preference for simplicity and expediency here. An attempt to include the entire `Project` seemed substantially more complex than the approach now in `dev`, and mostly unneeded, so I went the this (`metadata`) route.

**Comment by nsheff** [comment]
> For Sample-to-Project link, I've implemented a workaround that includes the metadata section. Inclusion of the entire Project reignited the circularity/recursion issues during serialization that I experienced amid the removal of the link a couple weeks ago and that we anticipated in this discussion. This felt acceptable given the use case discussed and the previously mentioned preference for simplicity and expediency here. An attempt to include the entire Project seemed substantially more complex than the approach now in dev, and mostly unneeded, so I went the this (metadata) route.

Ok sounds good.

**Comment by vreuter** [comment]
I also left in a commented-out section with an explanation of a slight implementation change that could be made depending on preferred interaction on the usage side. As it is now, linked attributes are accessed via something like  `sample.prj.output_dir`, but it's certainly possible to make them available directly, like `sample.output_dir`. I went this route since that sounded more like what may be in place in existing pipeline and is the current desire. I'm happy to change it around if the other way is preferred. Maybe @afrendeiro has a thought/preference.
